File: lib/sftp.ex
defmodule Sftp do
  use GenServer
  require Logger
  import Cfg
  import Notification

  alias Database.Queue
  alias Database.History

  alias :ssh, as: SSH
  alias :ssh_sftp, as: SFTP
  alias :timer, as: Timer

  ## Client API
  def start_link(opts \\ []) do
    GenServer.start_link(__MODULE__, :ok, [name: __MODULE__] ++ opts)
  end

  @spec add() :: :ok
  def add do
    GenServer.cast(__MODULE__, :add)
  end

  @spec do_exception() :: any()
  def do_exception do
    GenServer.call(__MODULE__, :do_exception, :infinity)
  end

  @spec launch_interval_check() :: {:error, any()} | {:ok, :timer.tref()}
  def launch_interval_check do
    Logger.info("Starting queue check with check interval: #{interval()}ms")
    Timer.apply_interval(interval(), Sftp, :add, [])
  end

  ## Callbacks (Server API)
  @spec init(:ok) :: {:ok, pid()}
  def init(:ok) do
    Logger.info("Starting Sftp module")
    SSH.start()

    case launch_interval_check() do
      {:ok, pid} ->
        Logger.debug("Internal check spawned: #{inspect(pid)}")

      {:error, error} ->
        Logger.error("Internal check spawn failed with error: #{inspect(error)}")
    end

    {:ok, self()}
  end

  defp sftp_open_and_process_upload(
         connection,
         local_file,
         remote_handle,
         remote_dest_file,
         channel
       ) do
    case SFTP.open(channel, remote_dest_file, [:write], sftp_open_channel_timeout()) do
      {:ok, handle} ->
        local_size = Utils.local_file_size(local_file)
        remote_size = Utils.remote_file_size(remote_handle)

        Logger.debug(
          "Local file: #{local_file} (#{local_size}); Remote file: #{remote_dest_file} (#{remote_size})"
        )

        cond do
          remote_size <= 0 ->
            Logger.info("Found an empty remote file. Uploading file to remote")
            Utils.stream_file_to_remote(channel, handle, local_file, local_size)

          remote_size != local_size ->
            Logger.info("Local and remote files are different. Uploading file to remote")
            Utils.stream_file_to_remote(channel, handle, local_file, local_size)

          remote_size == local_size ->
            Logger.info("Found file of same size already uploaded. Skipping")
        end

        Logger.debug("Closing file handle, channel and ssh connection")
        SFTP.close(channel, handle)
        SFTP.stop_channel(channel)
        SSH.close(connection)

      {:error, err} ->
        notification("Error opening write handle of remote file: #{inspect(err)}", :error)
    end
  end

  defp process_ssh_connection(connection, local_file, remote_dest_file) do
    Logger.debug(
      "Starting ssh channel for connection: #{inspect(connection)} for local file: #{local_file}"
    )

    case SFTP.start_channel(connection,
           blocking: false,
           pull_interval: 20,
           timeout: sftp_start_channel_timeout()
         ) do
      {:ok, channel} ->
        remote_dest_file = remote_dest_file |> String.to_charlist()
        remote_handle = SFTP.read_file_info(channel, remote_dest_file)
        Logger.debug("Started channel: #{inspect(channel)} for file: #{remote_dest_file}")

        connection
        |> sftp_open_and_process_upload(local_file, remote_handle, remote_dest_file, channel)

      {:error, err} ->
        notification("Error creating SFTP channel: #{inspect(err)}!", :error)
    end
  end

  def handle_cast(:add, _) do
    queue = DB.get_queue()

    unless Enum.empty?(queue) do
      build_clipboard()

      queue
      |> Enum.map(fn element ->
        element |> process_element
      end)
    end

    {:noreply, self()}
  end

  def handle_cast({:send_file, local_file, remote_dest_file}, _) do
    case SSH.connect(
           String.to_charlist(config()[:hostname]),
           config()[:ssh_port],
           ssh_opts(),
           ssh_connection_timeout()
         ) do
      {:ok, connection} ->
        Logger.debug("Processing connection with pid: #{inspect(connection)}")

        time =
          Timer.tc(fn ->
            connection |> process_ssh_connection(local_file, remote_dest_file)
          end)

        case time do
          {elapsed, _} ->
            Logger.debug("process_ssh_connection finished in: #{elapsed / 1000}ms")
        end

      {:error, cause} ->
        Logger.error("Error caused by: #{inspect(cause)}")
    end

    {:noreply, self()}
  end

  defp process_element(element) do
    case element do
      %Queue{local_file: local_file, remote_file: remote_dest_file, uuid: random_uuid} ->
        a_queue = %Queue{
          local_file: local_file,
          remote_file: remote_dest_file,
          uuid: random_uuid
        }

        if File.exists?(local_file) and File.regular?(local_file) and
             not Regex.match?(~r/.*-[a-zA-Z]{4,}$/, local_file) do
          local_file |> send_file(remote_dest_file <> Utils.file_extension(local_file))
          a_queue |> add_to_history |> DB.remove_from_queue()
        else
          Logger.debug("Local file not found or not a regular file: #{local_file}. Skipping.")
          a_queue |> DB.remove_from_queue()
        end

      :empty ->
        Logger.info("Empty queue. Ignoring request")
    end
  end

  @spec send_file(local_file :: String.t(), remote_dest_file :: String.t()) :: any
  def send_file(local_file, remote_dest_file) do
    GenServer.cast(__MODULE__, {:send_file, local_file, remote_dest_file})
  end

  @doc """
  Adds clipboard items to persistent history
  """
  @spec add_to_history(Queue.t()) :: Queue.t()
  def add_to_history(queue) do
    to_history = config()[:address] <> queue.uuid <> Utils.file_extension(queue.local_file)

    a_history = %History{
      content: to_history,
      timestamp: Timestamp.now_unix(),
      file: queue.local_file,
      uuid: UUID.uuid4(:hex)
    }

    count =
      DB.get_history()
      |> Enum.filter(fn element ->
        String.contains?(element.content, to_history)
      end)
      |> Enum.count()

    if count == 0, do: DB.add_history(a_history)

    # pass queue structure further:
    queue
  end

  @spec create_queue_string(Enum.t()) :: String.t()
  def create_queue_string(collection) when is_list(collection) do
    Enum.map(collection, fn an_elem ->
      case an_elem do
        %Queue{local_file: file_path, remote_file: _, uuid: uuid} ->
          config()[:address] <> uuid <> Utils.file_extension(file_path)

        # :empty
        _ ->
          ""
      end
    end)
    |> Enum.join("\n")
  end

  def create_queue_string(_), do: ""

  @doc """
  Creates content which will be copied to clipboard as http links
  """
  @spec build_clipboard() :: :ok
  def build_clipboard do
    clip_time =
      Timer.tc(fn ->
        DB.get_queue()
        |> create_queue_string
        |> Clipboard.put()
      end)

    case clip_time do
      {elapsed, _} ->
        Logger.debug("Clipboard routine done in: #{elapsed / 1000}ms")
        :ok
    end
  end

  def handle_call(:do_exception, _from, _ssh_connection) do
    raise "An exception!"
  end
end

File: src/sftp.rs
use crate::{
    config::AppConfig,
    database::{Database, History, QueueItem},
    notification::notification,
    utils, *,
};
use anyhow::{Context, Result};
use ssh2::Session;
use std::{fs::File, io::BufReader, net::TcpStream, path::Path, sync::Arc, time::Duration};
use tokio::{sync::mpsc, time};

#[derive(Debug)]
pub struct SftpManager {
    config: Arc<AppConfig>,
    database: Arc<Database>,
}


impl SftpManager {
    pub fn new(
        config: Arc<AppConfig>,
        database: Arc<Database>,
    ) -> (Self, mpsc::UnboundedReceiver<()>) {
        let (_tx, rx) = mpsc::unbounded_channel();
        (
            SftpManager {
                config,
                database,
            },
            rx,
        )
    }


    pub async fn start(self: Arc<Self>) {
        let mut interval =
            time::interval(Duration::from_millis(self.config.fs_check_interval));

        info!(
            "Starting SFTP queue processor with check interval: {}ms",
            self.config.fs_check_interval
        );

        loop {
            interval.tick().await;
            if let Err(e) = self.process_queue().await {
                error!("Error processing queue: {e:?}");
            }
        }
    }


    async fn process_queue(&self) -> Result<()> {
        let queue = self.database.get_queue()?;
        if !queue.is_empty() {
            // Build clipboard content
            self.build_clipboard(&queue)?;

            // Process each queue item
            for item in &queue {
                if let Err(e) = self.process_element(item).await {
                    error!("Error processing queue element: {e:?}");
                }
            }
        }
        Ok(())
    }


    fn build_clipboard(&self, queue: &[QueueItem]) -> Result<()> {
        let links: Vec<String> = queue
            .iter()
            .map(|item| {
                let config = self
                    .config
                    .select_config()
                    .expect("One of configs should always be selected!");
                format!(
                    "{}{}{}",
                    config.address,
                    item.uuid,
                    file_extension(&item.local_file)
                )
            })
            .collect();

        let content = links.join(", ");
        put_to_clipboard(&content)?;

        if self.config.notifications.clipboard {
            let _ = notification(
                "Link copied to clipboard",
                "clipboard",
                &self.config.notifications,
                &self.config.sounds,
            );
        }

        Ok(())
    }


    async fn process_element(&self, item: &QueueItem) -> Result<()> {
        let path = Path::new(&item.local_file);

        // Check if file exists and is regular
        if !path.exists() || !path.is_file() {
            debug!(
                "Local file not found or not a regular file: {}. Skipping.",
                item.local_file
            );
            self.database.remove_from_queue(&item.uuid)?;
            return Ok(());
        }

        // Check for temp file pattern
        if TEMP_PATTERN.is_match(&item.local_file) {
            debug!("File matches temp pattern, skipping: {}", item.local_file);
            self.database.remove_from_queue(&item.uuid)?;
            return Ok(());
        }

        // Upload file
        let remote_file = format!("{}{}", item.remote_file, file_extension(&item.local_file));
        self.send_file(&item.local_file, &remote_file).await?;

        // Add to history
        self.add_to_history(item)?;

        // Remove from queue
        self.database.remove_from_queue(&item.uuid)?;

        Ok(())
    }


    async fn send_file(&self, local_file: &str, remote_file: &str) -> Result<()> {
        let config = &self
            .config
            .select_config()
            .expect("One of configs should always be selected!");

        // Connect via SSH
        let tcp = TcpStream::connect(format!("{}:{}", config.hostname, config.ssh_port))
            .context("Failed to connect to SSH server")?;
        tcp.set_read_timeout(Some(Duration::from_millis(
            self.config.ssh_connection_timeout,
        )))?;
        tcp.set_write_timeout(Some(Duration::from_millis(
            self.config.ssh_connection_timeout,
        )))?;

        let mut sess = Session::new()?;
        sess.set_tcp_stream(tcp);
        sess.handshake()?;

        // Authenticate
        let ssh_private_key = if config.ssh_key.is_empty() {
            ".ssh/id_ed25519"
        } else {
            &config.ssh_key
        };

        sess.userauth_pubkey_file(
            &config.username,
            None,
            Path::new(&home::home_dir().expect("Home dir has to be set!"))
                .join(ssh_private_key)
                .as_path(),
            if config.ssh_key_pass.is_empty() {
                None
            } else {
                Some(&config.ssh_key_pass)
            },
        )?;

        if !sess.authenticated() {
            anyhow::bail!("SSH authentication failed");
        }

        debug!("SSH connection established");

        // Start SFTP session
        let sftp = sess.sftp()?;
        debug!("SFTP session started");

        // Check remote file
        let local_size = local_file_size(local_file)?;
        let remote_size = sftp
            .stat(Path::new(remote_file))
            .map(|stat| stat.size.unwrap_or(0))
            .unwrap_or(0);

        debug!(
            "Local file: {local_file} ({local_size}); Remote file: {remote_file} ({remote_size})"
        );

        if remote_size > 0 && remote_size == local_size {
            info!("Found file of same size already uploaded. Skipping");
            return Ok(());
        }

        // Upload file
        let mut local = BufReader::new(File::open(local_file)?);
        let mut remote = sftp.create(Path::new(remote_file))?;

        stream_file_to_remote(
            &mut local,
            &mut remote,
            self.config.sftp_buffer_size,
            local_size,
        )?;

        if self.config.notifications.upload {
            let _ = notification(
                "Uploaded successfully.",
                "upload",
                &self.config.notifications,
                &self.config.sounds,
            );
        }
        Ok(())
    }


    fn add_to_history(&self, queue_item: &QueueItem) -> Result<()> {
        let config = self
            .config
            .select_config()
            .expect("One of configs should always be selected!");
        let content = format!(
            "{}{}{}",
            config.address,
            queue_item.uuid,
            file_extension(&queue_item.local_file)
        );

        // Check if already in history
        let history = self.database.get_history(None)?;
        let exists = history.iter().any(|h| h.content.contains(&content));

        if !exists {
            let history_item = History {
                content,
                timestamp: chrono::Local::now().timestamp(),
                file: queue_item.local_file.clone(),
                uuid: uuid::Uuid::new_v4().to_string(),
            };
            self.database.add_history(&history_item)?;
        }

        Ok(())
    }
}
